site/blog/unlisted/subqueries.md

```diff
@@ -0,0 +1,86 @@
+
+Subqueries also offer increased flexibility compared to joins. With subqueries, you have more control over the order in which the data is retrieved, and you can easily modify the query to return the data you need. Joins are often more rigid, and making changes to the query can be more complex and time-consuming. In PostgreSQL, subqueries can be used to retrieve data from a single table and then join the results with another table, eliminating the need for complex join conditions.
+
+Let's take a look at an example to see the difference between using joins and subqueries in PostgreSQL. Let's say we have two tables, “employee” and “department”. We want to retrieve the names of employees who work in a specific department.
+
+Using Joins:
+
+```sql
+SELECT employee.name
+FROM employee
+INNER JOIN department
+    ON employee.department_id = department.id
+WHERE department.name = 'Engineering';
+```
+
+Using Subqueries:
+
+```sql
+SELECT name
+FROM employee
+WHERE department_id = (
+    SELECT id
+    FROM department
+    WHERE name = 'Engineering'
+);
+```
+
+The subquery is simpler and easier to understand than the join query in this case. The subquery allows us to retrieve the department ID for the specified department, and then use it to filter the employees who work in that department.
+
+### Simplified Maintenance and Composability
+
+Using subqueries instead of joins can simplify maintenance. Because subqueries are easier to understand and modify, it is easier to make changes to your database without affecting other parts of your application. It is possible to compose a top-level query by importing and using queries defined in other source files and including them as subqueries via templating. Improvements or schema changes that affect the subqueries then are encapsulated away from the logic of the high-level query and other subqueries within the top-level query. In contrast, changes to joins can have a significant impact on other parts of your code and can be more difficult to maintain over time. To expand on this using the example above, if the department where you want to retrieve the employee names is split into two separate departments, but both departments are still joined to the same category in the "cost center" table the query above looks as follows:
+
+Using Joins:
+
+```sql
+SELECT employee.name
+FROM employee
+INNER JOIN cost_center
+    ON employee.cost_center_id = cost_center.id
+WHERE cost_center.name = 'R&D';
+```
+
+The query had to change completely which seems negligible at this scale but adds up for complex queries. On the other hand, using subqueries requires fewer changes:
```

Comment:
hmm maybe this is not the best example since the query seems the same as before but using another table?